_posts/2022-11-29-November-Update-Exploring-V3D.md

```diff
@@ -0,0 +1,67 @@
+
+# Working with the Raspberry Pi 4
+
+---
+
+Although GSoC was a great learning experience, as I mentioned in my last blog post, I was willing to learn more about the Linux graphics stack. Fortunately, when I started the Igalia CE, [Melissa Wen](https://melissawen.github.io/) pitched me a project to increase IGT test coverage on DRM/V3D kernel driver. I was pretty glad to hear about the project as it allowed me to learn more about how a GPU works.
+
+## The Project
+
+Currently, V3D only has three basic IGT tests: `v3d_get_bo_offset`, `v3d_get_param`, and `v3d_mmap`. So, the basic goal of my CE project was to add more tests to the V3D driver.
+
+As the general DRM-core tests were in a good shape on the V3D driver, I started to think together with my mentors about more driver-specific tests for the driver.
+
+By checking the V3D UAPI, you can see that the V3D has eleven ioctls, so there is yet a lot to test for the V3D on IGT.
+
+First, there are Buffer Object (BO) related-ioctls: `v3d_create_bo`, `v3d_wait_bo`, `v3d_mmap_bo`, and `v3d_get_bo_offset`. The Buffer Objects are shared-memory objects that are used for storage on the GPU and testing them is important to make sure that memory is being correctly allocated. Different from the VC4, the V3D has an MMU between the GPU and the bus, allowing it to not allocate objects contiguously. Therefore, the idea was to develop tests for `v3d_create_bo` and `v3d_wait_bo`.
+
+Next, there are Performance Monitor (perfmon) related-ioctls: `v3d_perfmon_create`, `v3d_perfmon_destroy`, and `v3d_perfmon_get_values`. Performance Monitors are basically registers that are used for monitoring the performance of the V3D system. So, tests were designed to assure that the driver was creating perfmons properly and was resilient to incorrect requests, such as trying to get a value from a non-existent perfmon.
+
+And finally, the most interesting type of ioctls: the job submission ioctls. You can use the `v3d_submit_cl` ioctl to submit commands to the 3D engine, which is a tiled engine. When I think about tiled rendering, I always think about a [Super Nintendo](https://www.youtube.com/@RGMechEx), but things can get a bit more complicated than a SNES as you can see [here](https://developer.samsung.com/galaxy-gamedev/resources/articles/gpu-framebuffer.html). The 3D engine is composed of a bin and render pipelines, each has its command list and the render depends on the binner. By testing the v3d_submit_cl ioctl, it is possible to test syncing between jobs and also the [V3D multisync ability](https://melissawen.github.io/blog/2022/05/10/multisync-p1).
```

Comment:
instead of saying `and the render depends on the binner` I think it's better to explain/describe what is done in the bin and render stage/pipeline.